# Homework

**_NOTE:_** sometimes your answer doesn't match one of the options exactly. That's fine. Select the option that's closest to your solution.

## Dataset

In this homework, we will use the lead scoring dataset Bank Marketing dataset. Download it from here.

Or you can do it with wget:

```bash
wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/course_lead_scoring.csv
```

In this dataset our desired target for classification task will be converted variable - has the client signed up to the platform or not.

## Data preparation

- Check if the missing values are presented in the features.
- If there are missing values:
  - For caterogiral features, replace them with 'NA'
  - For numerical features, replace with with 0.0

In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt

In [2]:
data = 'https://raw.githubusercontent.com/alexeygrigorev/datasets/master/course_lead_scoring.csv'

In [3]:
!wget $data -O course_lead_scoring.csv

--2025-10-16 20:18:45--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/course_lead_scoring.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.110.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 80876 (79K) [text/plain]
Saving to: ‘course_lead_scoring.csv’

course_lead_scoring 100%[===================>]  78.98K  --.-KB/s    in 0.02s   

2025-10-16 20:18:45 (3.72 MB/s) - ‘course_lead_scoring.csv’ saved [80876/80876]



In [4]:
df = pd.read_csv('course_lead_scoring.csv')


In [5]:
df.dtypes

lead_source                  object
industry                     object
number_of_courses_viewed      int64
annual_income               float64
employment_status            object
location                     object
interaction_count             int64
lead_score                  float64
converted                     int64
dtype: object

In [6]:
numerical = ['number_of_courses_viewed', 'annual_income', 'interaction_count', 'lead_score']
numerical = df.select_dtypes(exclude='object').columns.drop('converted')
numerical = df.select_dtypes(exclude='object').columns.drop('converted')
numerical

Index(['number_of_courses_viewed', 'annual_income', 'interaction_count',
       'lead_score'],
      dtype='object')

In [7]:
categorical = df.select_dtypes(include=['object']).columns

In [8]:
df[numerical] = df[numerical].fillna(0.0)
df[categorical] = df[categorical].fillna('NA')

# Question 1

What is the most frequent observation (mode) for the column industry?

- `NA`
- `technology`
- `healthcare`
- `retail`

In [9]:
df.industry.mode()

0    retail
Name: industry, dtype: object

# Question 2

Create the correlation matrix for the numerical features of your dataset. In a correlation matrix, you compute the correlation coefficient between every pair of features.

What are the two features that have the biggest correlation?

- `interaction_count` and `lead_score`
- `number_of_courses_viewed` and `lead_score`
- `number_of_courses_viewed` and `interaction_count`
- `annual_income` and `interaction_count`

Only consider the pairs above when answering this question.
Split the data

- Split your data in train/val/test sets with 60%/20%/20% distribution.
- Use Scikit-Learn for that (the `train_test_split` function) and set the seed to ''42''.
- Make sure that the target value `y` is not in your dataframe.

In [10]:
df[numerical].corr()

number_of_courses_viewed  annual_income  \
number_of_courses_viewed                  1.000000       0.009770   
annual_income                             0.009770       1.000000   
interaction_count                        -0.023565       0.027036   
lead_score                               -0.004879       0.015610   

                          interaction_count  lead_score  
number_of_courses_viewed          -0.023565   -0.004879  
annual_income                      0.027036    0.015610  
interaction_count                  1.000000    0.009888  
lead_score                         0.009888    1.000000

interaction_count and lead_score: 0.009888, number_of_courses_viewed and lead_score: -0.004879

number_of_courses_viewed and interaction_count: -0.023565


annual_income and interaction_count: 0.027036


In [11]:
from sklearn.model_selection import train_test_split

In [12]:
df_train_full, df_test = train_test_split(df.drop(columns='converted'), test_size=0.2, random_state=42)

In [13]:
df_train, df_val = train_test_split(df_train_full, test_size=0.25, random_state=42)

In [14]:
df_train_full

lead_source       industry  number_of_courses_viewed  annual_income  \
1066  social_media  manufacturing                         2        44403.0   
638         events         retail                         3        38048.0   
799   social_media      education                         2        71399.0   
380       referral      education                         2        47912.0   
303       paid_ads     healthcare                         1        34806.0   
...            ...            ...                       ...            ...   
1130        events  manufacturing                         1        57039.0   
1294        events     healthcare                         2        56185.0   
860       paid_ads  manufacturing                         1        56402.0   
1459      paid_ads     technology                         1        45688.0   
1126        events  manufacturing                         0        43862.0   

     employment_status       location  interaction_count  lead_score  
1066     self_employed      australia                  1        0.71  
638            student  north_america                  6        0.97  
799                 NA         europe                  1        0.51  
380           employed      australia                  1        0.04  
303           employed         europe                  4        0.32  
...                ...            ...                ...         ...  
1130          employed  south_america                  2        0.30  
1294           student  south_america                  2        0.44  
860            student  north_america                  1        0.02  
1459           student  north_america                  3        0.02  
1126     self_employed           asia                  4        0.21  

[1169 rows x 8 columns]

In [15]:
# Reset Index für saubere Zeilen
df_train = df_train.reset_index(drop=True)
df_val = df_val.reset_index(drop=True)
df_test = df_test.reset_index(drop=True)

In [16]:
y_train = df_train.converted.values
y_val = df_val.converted.values
y_test = df_test.converted.values

AttributeError: 'DataFrame' object has no attribute 'converted'

In [ ]:
df_train_full = df_train_full.reset_index(drop=True)

# Question 3

- Calculate the mutual information score between y and other categorical variables in the dataset. Use the training set only.
- Round the scores to 2 decimals using round(score, 2).

Which of these variables has the biggest mutual information score?

- `industry`
- `location`
- `lead_source`
- `employment_status`

In [ ]:
from sklearn.metrics import mutual_info_score

In [ ]:
df_train_full, df_test = train_test_split(df, test_size=0.2, random_state=42)

In [ ]:
df_train, df_val = train_test_split(df_train_full, test_size=293, random_state=42)

In [ ]:
def calculate_mi(series):
    return mutual_info_score(series, df_train.converted)

df_mi = df_train[categorical].apply(calculate_mi)
df_mi = df_mi.sort_values(ascending=False).to_frame(name='MI')


df_mi.round(2)

lead_source

In [ ]:
#df_train_full, df_test = train_test_split(df.drop(columns='converted'), test_size=0.2, random_state=42)

# Question 4

- Now let's train a logistic regression.
- Remember that we have several categorical variables in the dataset. Include them using one-hot encoding.
- Fit the model on the training dataset.
- To make sure the results are reproducible across different versions of Scikit-Learn, fit the model with these parameters:
- model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
- Calculate the accuracy on the validation dataset and round it to 2 decimal digits.

What accuracy did you get?

- 0.64
- 0.74
- 0.84
- 0.94

In [ ]:
## One-hot encodingabs

In [ ]:
from sklearn.feature_extraction import DictVectorizer

In [ ]:
df_train, df_val = train_test_split(df_train_full, test_size=293, random_state=42)

In [ ]:
df_train

In [ ]:
categorical = list(categorical)
numerical = list(numerical)

In [ ]:
train_dict = df_train[categorical + numerical].to_dict(orient='records')
train_dict[0]

In [ ]:
dv = DictVectorizer(sparse=False)
train_dict = df_train[categorical + numerical].to_dict(orient='records')
X_train = dv.fit_transform(train_dict)

val_dict = df_val[categorical + numerical].to_dict(orient='records')
X_val = dv.transform(val_dict)

In [ ]:
X_train = dv.fit_transform(train_dict)

In [ ]:
X_train.shape

In [ ]:
dv.get_feature_names_out()

In [ ]:
y_train = df_train.converted.values
y_val = df_val.converted.values

In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
model.fit(X_train, y_train)

In [ ]:
model.predict_proba(X_val)

In [ ]:
y_pred = model.predict_proba(X_val)[:, 1]

In [ ]:
converted = y_pred >= 0.5

In [ ]:
y_pred.mean()

In [ ]:
(y_val == converted).mean()

In [ ]:
y_pred = model.predict(X_val)

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [ ]:
round(accuracy_score(y_val, y_pred),2)

In [ ]:
dicts_train_full = df_train_full[categorical + numerical].to_dict(orient='records')

In [ ]:
dv = DictVectorizer(sparse=False)

In [ ]:
X_train_full = dv.fit_transform(dicts_train_full)

In [ ]:
y_train_full = df_train_full.converted.values

In [ ]:
model_baseline = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
model_baseline.fit(X_train, y_train)

In [ ]:
model.intercept_[0]

In [ ]:
model.coef_[0].round(3)

In [ ]:
dicts_test = df_test[categorical + numerical].to_dict(orient='records')

In [ ]:
X_test = dv.transform(dicts_test)

In [ ]:
y_pred = model.predict_proba(X_train)[:, 1]

In [ ]:
converted_decision = (y_pred >= 0.5)
converted_decision

In [ ]:
#(y_val == converted_decision).mean()

In [ ]:
y_pred = model.predict(X_val)

In [ ]:
accuracy_score(y_val, y_pred)

In [ ]:
ac = accuracy_score(y_val, y_pred)

In [ ]:
y_val_pred_baseline = model_baseline.predict(X_val)
accuracy_baseline = accuracy_score(y_val, y_val_pred_baseline)
print(f'Accuracy: {accuracy_baseline:.2f}')

In [ ]:
round(ac,2)

In [ ]:
# A function to evaluate the model without a specific feature
def evaluate_without_feature(feature_to_exclude):
    # Get the feature names from the DictVectorizer
    feature_names = dv.get_feature_names_out()
    
    # Find indices of features to keep (all except the ones to exclude)
    exclude_indices = [i for i, feat in enumerate(feature_names) if feature_to_exclude in feat]
    keep_indices = [i for i in range(len(feature_names)) if i not in exclude_indices]
    
    # Create new datasets without the feature
    X_train_reduced = X_train[:, keep_indices]
    X_val_reduced = X_val[:, keep_indices]
    
    # Train the model
    model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
    model.fit(X_train_reduced, y_train)
    
    # Evaluate on validation set
    y_val_pred = model.predict(X_val_reduced)
    accuracy = accuracy_score(y_val, y_val_pred)
    
    return accuracy

# Question 5

- Let's find the least useful feature using the feature elimination technique.
- Train a model using the same features and parameters as in Q4 (without rounding).
- Now exclude each feature from this set and train a model without it. Record the accuracy for each model.
- For each feature, calculate the difference between the original accuracy and the accuracy without the feature.

Which of following feature has the smallest difference?

- `'industry'`
- `'employment_status'`
- `'lead_score'`

**_NOTE:_** The difference doesn't have to be positive.

numerical = ['number_of_courses_viewed', 'annual_income', 'interaction_count', 'lead_score']
categorical = ['lead_source', 'industry', 'employment_status', 'location']

In [ ]:
df[['lead_source', 'converted']]

In [ ]:
#df_train_full, df_test = train_test_split(df.drop(columns=['converted','lead_source']), test_size=0.2, random_state=42)

In [ ]:
X = df[numerical + categorical]
y = df['converted']

In [ ]:
df_train_full, df_test = train_test_split(df, test_size=0.2, random_state=42)

In [ ]:
df_train, df_val = train_test_split(df_train_full, test_size=293, random_state=42)

In [ ]:
dicts_train_full = df_train_full[categorical + numerical].to_dict(orient='records')

In [ ]:
dv = DictVectorizer(sparse=False)
X_train_full = dv.fit_transform(dicts_train_full)

In [ ]:
y_train_full = df_train_full.converted.values

In [ ]:
dicts_test = df_test[categorical + numerical].to_dict(orient='records')

In [ ]:
# Evaluate the model without each feature
features_to_test = ['industry', 'employment_status', 'lead_score']
results = {}

for feature in features_to_test:
    accuracy_without_feature = evaluate_without_feature(feature)
    difference = accuracy_baseline - accuracy_without_feature
    results[feature] = {
        'accuracy': accuracy_without_feature,
        'difference': difference
    }
    print(f"Without {feature}: Accuracy = {accuracy_without_feature:.6f}, Difference = {difference:.6f}")

# Find the feature with the smallest difference
min_feature = min(results, key=lambda x: abs(results[x]['difference']))
print(f"\nFeature with smallest impact: {min_feature}, Difference: {results[min_feature]['difference']:.6f}")

# Question 6

- Now let's train a regularized logistic regression.
- Let's try the following values of the parameter `C: [0.01, 0.1, 1, 10, 100]`.
- Train models using all the features as in Q4.
- Calculate the accuracy on the validation dataset and round it to 3 decimal digits.

Which of these `C` leads to the best accuracy on the validation set?

- 0.01
- 0.1
- 1
- 10
- 100

**_NOTE:_** If there are multiple options, select the smallest C.

In [ ]:

c_values = [0.01, 0.1, 1, 10, 100]

# Dictionary to store results
c_results = {}

# Train and evaluate models for each C value
for c in c_values:
    # Train model with current C value
    model = LogisticRegression(solver='liblinear', C=c, max_iter=1000, random_state=42)
    model.fit(X_train, y_train)
    
    # Evaluate on validation set
    y_val_pred = model.predict(X_val)
    accuracy = accuracy_score(y_val, y_val_pred)
    
    # Store result
    c_results[c] = round(accuracy, 3)
    
    print(f"C = {c}, Validation Accuracy = {c_results[c]}")

best_accuracy = max(c_results.values())
best_c_values = [c for c, acc in c_results.items() if acc == best_accuracy]
best_c = min(best_c_values)

print(f"\nBest C value: {best_c} with validation accuracy: {c_results[best_c]}")

# Submit the results

- Submit your results here: https://courses.datatalks.club/ml-zoomcamp-2025/homework/hw03
- If your answer doesn't match options exactly, select the closest one